core/includes/install.inc

```diff
@@ -454,53 +456,29 @@ abstract class DatabaseTasks {
-      if ($original_connection_info['default']['host'] === 'localhost') {
+      $localhost_invalid = $original_connection_info['default']['host'] === 'localhost';
+      $local_ip_invalid = $original_connection_info['default']['host'] === '127.0.0.1';
+      if ($localhost_invalid) {
         $modified_connection_info['default']['host'] = '127.0.0.1';
         $modified_connection_info['default']['port'] = '3306';
       }
-      elseif ($original_connection_info['default']['host'] === '127.0.0.1') {
+      elseif ($local_ip_invalid) {
         $modified_connection_info['default']['host'] = 'localhost';
         $modified_connection_info['default']['port'] = '';
       }
       if ($original_connection_info != $modified_connection_info) {
         Database::removeConnection('default');
         Database::addConnectionInfo('default', 'default', $modified_connection_info['default']);
         try {
-          Database::getConnection();
-          $last_exception = FALSE;
-
-          // Connection was successful, but settings.php needs to be updated.
-          $db_settings = $modified_connection_info['default'];
-          $settings['database'] = array(
-            'value' => array(
-              'database' => $db_settings['database'],
-              'username' => $db_settings['username'],
-              'password' => $db_settings['password'],
-              'host' => $db_settings['host'],
-            ),
-            'required' => TRUE,
-          );
-          // Include the driver if it is not 'mysql'.
-          if ($db_settings['driver'] != 'mysql') {
-            $settings['database']['value']['driver'] = $db_settings['driver'];
-          }
-          // Include the prefix if one has been set.
-          if (!empty($db_settings['prefix'])) {
-            $settings['database']['value']['prefix'] = $db_settings['prefix'];
-          }
-          // Include the port if the default has been overridden.
-          if ($db_settings['port']) {
-            $settings['database']['value']['port'] = $db_settings['port'];
-          }
-          // Include the charset if the default has been overridden.
-          if ($db_settings['charset']) {
-            $settings['database']['value']['charset'] = $db_settings['charset'];
-          }
-          // Include the collation if the default has been overridden.
-          if ($db_settings['collation']) {
-            $settings['database']['value']['collation'] = $db_settings['collation'];
-          }
-          backdrop_rewrite_settings($settings);
+          Database::getConnection('default', 'default');
+
+          // The modified database connection was successful. Provide an
+          // informative help message asking the installer to adjust the
+          // values.
+          $extra_help = st('Connecting to "@broken_host" was not successful, but connecting to "@working_host" with the same credentials was successful. Try setting the database connection host to "@working_host".', array(
```

Comment:
The database connection setting isn't visible at this point. I suggest to mention that it is in the "Advanced options".

Reply:
I've clarified this in the text.